### ISSUE TITLE
Allow "direct" Docker host config

### ISSUE THREAD

**Issue by robnagler** [jupyterhub/dockerspawner#349]
docker.py often lags Go interface by quite some time. For example, there's support for `--gpus` in the CLI but not docker.py. I propose that we allow DockerSpawner to bypass HostConfig, and just send parameters directly to the Docker daemon. There's a hack solution to [support `gpus` in RSDockerSpawner](https://github.com/radiasoft/rsdockerspawner/blob/6f1653ce4c0cb9532d77c5f9c03b8c64a7ff051b/rsdockerspawner/rsdockerspawner.py#L100), which would be better implemented if DockerSpawner allowed `host_config` to be modified after the `create_host_config`.

I think the general docker.py takes on converting arguments is flawed, that is, HostConfig shouldn't be mapping Python names to Go config names. There are [63 parameters to HostConfig](https://github.com/docker/docker-py/blob/72f4f527ad60b2a677b883dc54669ebe98f0879f/docker/types/containers.py#L158). This is problematic for a number of reasons, mostly because we can't look at the Go code to figure out the names for things. It would be cleaner to just pass the Go parameters directly.

I'm happy to submit a PR, but this is something that obviously needs discussion. My hack works for now.

**Comment by samon11** [comment]
I second this feature request. I am struggling to have DockerSpawner start containers with the `--gpus` flag. @robnagler do you know of any workarounds in the meantime?

**Comment by robnagler** [comment]
@samon11 You will need to create a subclass of DockerSpawner. This can be done [in the config](https://github.com/jupyterhub/dockerspawner/issues/172#issuecomment-323580954), and then you need [to override the docker method](https://github.com/radiasoft/rsdockerspawner/blob/6f1653ce4c0cb9532d77c5f9c03b8c64a7ff051b/rsdockerspawner/rsdockerspawner.py#L100).

**Comment by minrk** [comment]
`extra_host_config` lets us pass anything through docker's hostconfig. There is an example for gpus in https://github.com/jupyterhub/dockerspawner/issues/244#issuecomment-785866378:

```python
import docker

c.DockerSpawner.extra_host_config = {
    "device_requests": [
        docker.types.DeviceRequest(
            count=-1,
            capabilities=[["gpu"]],
        ),
    ],
}
```

**Comment by acelere** [comment]
@minrk , could you confirm you were able to run TensorFlow with this setup?
It is very weird. The container seems to launch and accept the parameters to expose the GPU, but TF does not see it from within JupyterLab.
I tried what you suggested above, plus even adding "a few extra parameters to make sure", in a *desperate act* (everyone, this is just to illustrate the test I did, no need for extra parameters)

```python
import docker

c.DockerSpawner.extra_host_config = {
    "device_requests": [
        docker.types.DeviceRequest(
            count=-1,
            capabilities=[["gpu"]],
            driver="nvidia",
            options={"--gpus":"all"}
        ),
    ],
}

```
After logging into Jupyterhub, I get this from testing nvidia from a shell:
`nvidia-smi`
![nvidia_smi](https://user-images.githubusercontent.com/12059585/162530244-b3eafddf-5655-457a-a7b2-6316d6621dd9.png)

**BUT** when I try to get the GPU from TF, I get only the CPU:


![tensor_flow_test](https://user-images.githubusercontent.com/12059585/162530639-1eae653d-49d8-4cbb-8d49-630556690b29.png)

The container I am spawning this from was built with this Dockerfile:

```bash
FROM nvcr.io/nvidia/tensorflow:22.03-tf2-py3

ENV DEBIAN_FRONTEND=noninteractive
ENV TZ=Etc/UTC

RUN apt update && \
    apt-get -y install tzdata keyboard-configuration

RUN apt install -y zip \
    pip

RUN pip3 install --upgrade pip

RUN pip3 install \
    bqplot \
    control \
    ipywidgets \
    jax \
    jaxlib \
    jupyterhub \
    jupyterlab \
    matplotlib \
    nbconvert \
    notebook \
    numpy \
    pandas \
    traitlets

# create a user, since we don't want to run as root
RUN useradd -m jovyan && \
    usermod -a -G users,staff jovyan

ENV HOME=/home/jovyan
WORKDIR $HOME

RUN chown -R :users /usr/local/share/jupyter/lab
RUN chmod -R 775 /usr/local/share/jupyter/lab

USER jovyan

RUN mkdir /home/jovyan/work && \
    chown -R jovyan /home/jovyan/work

CMD ["jupyterhub-singleuser"]
```

Any ideas???

**Comment by acelere** [comment]
OK. Figured it out.
I ran the stock NVIDIA container interactively with:
` docker run --gpus all -it --rm nvcr.io/nvidia/tensorflow:22.03-tf2-py3`
and got the following message back:
![nvidia_01](https://user-images.githubusercontent.com/12059585/162813736-eee8bb17-f6eb-46c8-b8ae-fa5861c11fb7.PNG)

So I need a new GPU :)

Anyway, I guess the hint here is to run the container and check the errors...because not all of the output is presented to the user on Jupyterhub.

Cheers.
